added pxe nic support for redfish driver

kpradip001 opened the pull request:
Currently Ironic is not able to figure out PXE NICs available in the system using Redfish calls.
This PR makes sure that every-time when port is created, it checks BIOS and UEFI boot modes of node and fetch the pxe device interfaces values matching to nic id and then update the port's pxe_enabled field.

kpradip001 commented:
I have added PR for pxe nic support in redfish driver, testing is completed.
Currently working on adding unit test cases.

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
     specific incompatibilities and introduction of vendor value added
```
Can you remove lines 91-94 since we are now providing custom code here?

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,6 +43,47 @@
+    BOOT_MODE_MAP = {
+        sushy.BOOT_SOURCE_MODE_UEFI: boot_modes.UEFI,
+        sushy.BOOT_SOURCE_MODE_BIOS: boot_modes.LEGACY_BIOS
+    }
```
Can we just use the constants defined in the superclass instead of redefining them here?

cdearborn replied:
maybe redfish_inspect.CPU_ARCH_MAP?

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
+
+        """
+        node = task.node
+        system = redfish_utils.get_system(node)
```
It looks like a lot of the code below was copied from RedfishInspect.inspect_hardware().  Instead of copying all that code, could we just call: super(DracRedfishInspect, self).inspect_hardware()?  We could let it create the Ports, and then set pxe_enabled on the ports after.

kpradip001 replied:
Good idea! Done.

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
+                    pxe_dev_nics.append(mac_address)
+
+        elif system.boot.mode == 'bios':
+            redfish_username = task.node.driver_info.get('redfish_username')
```
Looks like we don't need to get redfish_username and redfish_password here since we already did above.

kpradip001 replied:
Done.

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
+                redfish_username, redfish_password))
+
+            if job_response.status_code != _JOB_RESPONSE_CODE:
+                LOG.debug("FAIL, status code not 202, code is: %(code), "
```
I would think we would want to raise an exception here instead of just returning an empty list for pxe_dev_nics

cdearborn replied:
maybe a HardwareInspectionFailure exception?

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
+            try:
+                job_id = re.search("JID_.+", job_id).group()
+            except AttributeError:
+                LOG.debug("FAIL: detailed error message: %(err)",
```
Same here

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
+                            pxe_dev_nics.append(mac_address)
+                    return pxe_dev_nics
+                elif current_time.seconds >= _TIMEOUT_SECONDS:
+                    LOG.debug("FAIL, Timeout of %(second) seconds has been"
```
Same here

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -44,7 +93,261 @@ class DracRedfishInspect(redfish_inspect.RedfishInspect):
+            start_time = datetime.now()
+            while True:
+                current_time = (datetime.now() - start_time)
+                xml_data = requests.get('%s%s%s' % (
```
Would you mind emailing me what you get back for xml_data?  I'm very curious to see what it is.  Thanks!

kpradip001 replied:
I sent you an email, Done.

kpradip001 commented:
> Hey Pradip, nice job figuring out how to get this working! I'm hoping that there is a way that we can call into the super class instead of reproducing all of that code here. That would help to reduce the size of this patch, plus reduce the maintenance burden if we had only 1 copy of the code. I would like to see HardwareInspectionFailure exceptions being raised when critical errors occur.

Hi Chris, thanks for review comments, I have addressed all the review comments in latest commit. please review it.

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +41,164 @@
+            enabled_macs = {nic_mac: nic_state
+                            for nic_mac, nic_state in macs.items()
+                            if nic_state == sushy.STATE_ENABLED}
+            if enabled_macs:
```
I think we could make this even simpler by removing the highlighted code and instead getting the ports for the node using:
objects.Port.list_by_node_id(task.context, node.id)
We could then iterate through those ports below and check to see if the mac address on the port is in pxe_dev_nics

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +41,164 @@
+                               'reached before marking the job completed.'),
+                             {'second': _TIMEOUT_SECONDS})
+                    raise exception.HardwareInspectionFailure(error=error)
+                    break
```
Do you think it would make sense to to add an else clause and sleep for 5 seconds or something like that?  If the job takes a while to finish, then we might want to do that.  If the job finishes very quickly, then I think the code is fine without the sleep.

kpradip001 replied:
Hi Chris, It takes only 1 to 5 seconds to finish the job. So, I think we don't need to add here time.sleep()

cdearborn replied:
sounds good.

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +41,164 @@
+                    auth=(redfish_username, redfish_password), verify=False)
+                if _SYSTEM_CONFIG_TAG in str(xml_data.__dict__):
+                    export_list = xml_data.__dict__.get('_content').split('\n')
+                    export_list = [line for line in export_list
```
It would be really nice if we could use the xml parsing library instead of just doing string parsing.  I think you can use xml.etree.ElementTree.fromstring() to do this.

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +41,164 @@
+                        redfish_ip, _MAC_URL, nic_id), auth=(
+                        redfish_username, redfish_password), verify=False)
+                    mac_address = json.loads(mac_response.__dict__.get(
+                        '_content')).get('MACAddress')
```
Could we use sushy here to get the MAC addresses?  See my comment below...

kpradip001 replied:
We can get MAC address from sushy.
But, In this case, we need to get the association of MAC address from it's nic_id, But, looks like it's association is not available in sushy.
So here I have used api to get the MAC address from it's nic_id.

cdearborn replied:
I see, thanks for looking into it!

kpradip001 commented:
> I think we could make the code simpler by getting the ports on the node in ironic instead of querying them from the iDRAC. Also, it would be nice to use sushy to get the MAC addresses where possible.

Hi @cdearborn ,
I have addressed your review comments,  Please take a look into this. 
Thanks.

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,154 @@
+                        redfish_ip, _MAC_URL, nic_id), auth=(
+                        redfish_username, redfish_password), verify=False)
+                    mac_address = json.loads(mac_response.__dict__.get(
+                        '_content')).get('MACAddress')
```
It would be great if you could create a _get_mac_address() method and call it in the 2 places in this method.

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,154 @@
+                            pxe_dev_nics.append(mac_address)
+                    return pxe_dev_nics
+                elif current_time.seconds >= _TIMEOUT_SECONDS:
+                    error = (_('FAIL, Timeout of %(second) seconds has been '
```
Can you change the error message to something like:
The job to export the system configuration did not complete within xxx seconds.

kpradip001 replied:
> Can you change the error message to something like:
> The job to export the system configuration did not complete within xxx seconds.

Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,154 @@
+            try:
+                job_id = re.search('JID_.+', job_id).group()
+            except drac_exceptions.BaseClientException as exc:
+                LOG.error('FAIL: detailed error message: %(err)',
```
Can you change this error message to something like:
Unable to find the status for job <job_id> in response: <err>

kpradip001 replied:
> Can you change this error message to something like:
> Unable to find the status for job <job_id> in response:

Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,154 @@
+                redfish_username, redfish_password))
+
+            if job_response.status_code != _JOB_RESPONSE_CODE:
+                error = (_('FAIL, status code not 202, code is: %(code), '
```
Can you change this error message to something like:
An error occurred when attempting to export the system configuration.  Status code: <code>.  Error: <error>

cdearborn replied:
Oh, could you also add a LOG.error(error) here too?

kpradip001 replied:
Done

kpradip001 commented:
> Nice job on this Pradip. Was wondering if you can change the error messages a little and factor the code to get the MAC address out into a separate method.

Hi Chris, I have updated PR by addressing above review comments, please take a look into this.
Will update PR with unit test cases as well soon.

kpradip001 commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,167 @@
+                    nic_id = system.bios.attributes[nic]
+                    # Get MAC address of the given nic_id
+                    mac_address = self._get_mac_address(
+                        nic_id, redfish_ip, redfish_username, redfish_password)
```
replace nic_id to [nic_id]
(will update in next commit)

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,167 @@
+        pxe_dev_nics = self._get_pxe_dev_nics(task)
+        if pxe_dev_nics is None:
+            LOG.warning("No PXE enabled NIC was found for node "
+                        "%(node_uuid)s.", {'node_uuid': node.uuid})
```
i think we have to "return states.MANAGEABLE" here.  Otherwise, we'll get a null pointer at line 98 below.

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,167 @@
+                if port.pxe_enabled != is_pxe_dev_nic:
+                    port.pxe_enabled = is_pxe_dev_nic
+                    port.save()
+                    LOG.info('Port updated with pxe enabled %(pxe)s '
```
Would you mind changing "pxe enabled" to "pxe_enabled" in the line above?

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,167 @@
+        """Get a list of pxe device interfaces.
+
+        :param task: a TaskManager instance.
+        :returns: Returns list of pxe device interfaces.
```
Maybe change "interfaces" in the line above to "MAC addresses"?

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,167 @@
+                               'did not complete within %(second) seconds'),
+                             {'second': _TIMEOUT_SECONDS})
+                    raise exception.HardwareInspectionFailure(error=error)
+                    break
```
don't need the break statement above since the raise will cause the loop to abort

kpradip001 replied:
Done

kpradip001 commented:
> Looking good! Had a few very minor comments, and I think we need to add a return to avoid a null pointer exception in one edge case.

Done

kpradip001 commented on ironic/tests/unit/db/utils.py:
```diff
@@ -425,6 +425,15 @@ def get_test_redfish_info():
 
+def get_test_drac_redfish_info():
+    return {
+        "redfish_address": "https://192.168.110.13",
```
At line no: 430 to 333, if we give correct credentials then only redfish api test cases executes successfully. 

And if we give dummy credentials (like line no: 421 to 424) then, while executing test cases consists redfish api call returns 401 or 404 response_code,
Please, let me know your though on this.

cdearborn replied:
Hey Rachit, this is because the code ends up calling _get_mac_address(), which uses requests to query the MAC address from the targeted server.  When you're running the test cases, it's actually retrieving the MAC from 110.13 over the network!
I think you should remove get_test_drac_redfish_info() and use get_test_redfish_info() instead.  In your test cases that result in the MAC being queried, you will either need to mock requests.get() or mock _get_mac_address() and write more test cases to test _get_mac_address() in isolation.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +387,81 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+    def setUp(self):
+        super(DracRedfishInspectionTestCase, self).setUp()
+        self.config(enabled_hardware_types=['idrac'],
+                    enabled_power_interfaces=['idrac-redfish'],
```
Instead of enabling the idrac-redfish version of the interfaces here, could we move this into BaseDracTest.setUp() and make this class inherit from BaseDracTest?

kpradip001 replied:
Hi Chris,
Pradip here,
I have inherited test_utils.BaseDracTest without any issue.
  
But while moving 'idrac-redfish' to BaseDracTest.setUp() ,getting attached error msg.


<img width="960" alt="errMsg" src="https://user-images.githubusercontent.com/34384282/67575382-1a304880-f75a-11e9-9d75-d0741f0447ea.png">

cdearborn replied:
Ah, I see.  It looks like it's trying to run the redfish test case against the wsman inspect implementation.  Let's leave it as-is then.

kpradip001 replied:
Okay

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +387,81 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+            pxe_dev_nics = task.driver.inspect._get_pxe_dev_nics(
+                task)
+            self.assertIsNone(pxe_dev_nics)
+            self.assertEqual(1, mock__get_pxe_dev_nics.call_count)
```
Can you add a test that validates that pxe_enabled is set correctly on the ports on the node?

kpradip001 replied:
Done

kpradip001 replied:
Done

kpradip001 commented:
> Hey Rachit, it looks like the _get_mac_address() method is actually querying the targeted node over the network for its MAC address. That's why the tests don't work with fake credentials. See my detailed comments.
> I think we need a test case that tests that pxe_enable is set correctly on the ports associated with the node. Also, can you add more test cases to test when exceptions are thrown in the code?

Hi Chris, Pradip here. (not Rachit :)  )

As you suggested, I have added below test cases:
i) BIOS/UEFI/NONE  boot modes with fake credentials,
ii) added test case for is pxe_enable is set (True / False) correctly
iii)timout_fail exception test case. etc

cdearborn commented:
My apologies Pradip!  I obviously didn't have enough coffee when I reviewed this!

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,168 @@
+                            mac_address = self._get_mac_address(
+                                nic_id, redfish_ip,
+                                redfish_username, redfish_password)
+                            pxe_dev_nics.append(mac_address)
```
The above lines depends on the firmware returning the MAC address in upper case.  Let's make it mac_address.upper() so that it won't matter if the firmware returns it in upper or lower case.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +388,176 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+                              task.driver.inspect._get_pxe_dev_nics, task)
+
+    @mock.patch.object(redfish_utils, 'get_system', autospec=True)
+    @mock.patch.object(DracRedfishInspect, '_get_mac_address', autospec=True)
```
I think you could remove the _get_mac_address mock, since I don't believe it will be called in this test.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +388,176 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+
+        with task_manager.acquire(self.context, self.node.uuid,
+                                  shared=True) as task:
+            task.driver.inspect.inspect_hardware(task)
```
Do we need to make the inspect_hardware() call?  I would think just calling _get_pxe_dev_nics() below would be all we want to do.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +388,176 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+
+    @mock.patch.object(DracRedfishInspect, '_get_pxe_dev_nics', autospec=True)
+    @mock.patch.object(redfish_utils, 'get_system', autospec=True)
+    def test_inspect_hardware_without_pxe_dev_nics(
```
This test mocks _get_pxe_dev_nics() above, but then for the test, calls into _get_pxe_dev_nics() below, so it isn't actually testing any of the code in inspect.py.  I think you want to call task.driver.inspect.inspect_hardware() on line 517 below instead of _get_pxe_dev_nics().

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +388,176 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+    @mock.patch.object(objects.Port, 'list_by_node_id', autospec=True)
+    @mock.patch.object(DracRedfishInspect, '_get_pxe_dev_nics', autospec=True)
+    @mock.patch.object(redfish_utils, 'get_system', autospec=True)
+    def test_inspect_hardware_with_set_port_pxe_enabled(
```
Very nice!  I love this test!

kpradip001 replied:
Thank you

cdearborn commented on ironic/drivers/modules/drac/inspect.py:
```diff
@@ -35,16 +42,168 @@
+            redfish_username, redfish_password), verify=False)
+        mac_address = json.loads(mac_response.__dict__.get(
+            '_content')).get('MACAddress')
+        return mac_address
```
Let's return mac_address.upper() here so if some versions of the firmware happen to return the MAC in lower case, it won't matter.  See line 99 above...

kpradip001 replied:
Done

kpradip001 commented:
> Looks really good. One thought that I had is that the code currently depends on the firmware returning the MAC address in upper case. Let's put in a call to .upper() so it won't matter if some firmware returns the MAC in lower case. Also had a few comments on the test cases.

Hi Chris,  I have addressed all the review comments, please review this.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +388,170 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+
+        with task_manager.acquire(self.context, self.node.uuid,
+                                  shared=True) as task:
+            task.driver.inspect.inspect_hardware(task)
```
I don't think the above line is needed since the line below is calling the method under test directly.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_inspect.py:
```diff
@@ -378,3 +388,170 @@ def test__get_pxe_dev_nics_with_empty_list(self, mock_get_drac_client):
+
+        with task_manager.acquire(self.context, self.node.uuid,
+                                  shared=True) as task:
+            task.driver.inspect.inspect_hardware(task)
```
I don't think the line above is needed since the line below is calling the _get_pxe_dev_nics() method directly.

kpradip001 replied:
Done

kpradip001 commented:
> I think we have a couple extra calls to inspect_hardware() in the test cases. Looks great other than that.

Thanks Chris, I have removed extra inspect_hardware call, please review this.

kpradip001 commented:
> Looks good! Please go ahead and submit this upstream to Ironic.

Thank you Chris :) ,
I have created patch in upstream ironic for Adding PXE nic support to ``idrac-redfish`` inspect interface in redfish OOB introspection
https://review.opendev.org/#/c/693328/

We can close this PR now.
Thanks again.

cdearborn commented:
Closing since this has been submitted upstream.
